pylabnet/utils/zi_hdawg_pulseblock_handler/zi_hdawg_pb_handler.py

```diff
@@ -0,0 +1,252 @@
+
+import numpy as np
+from pylabnet.utils.pulseblock.pb_sample import pb_sample
+
+
+# Sampling rate of HDWAG sequencer (300 MHZ)
+SEQ_SAMP_RATE = 300e6
+
+# Duration of setDIO() commands to be used as offset in wait() commands
+SETDIO_OFFSET = 4
+
+
+class DIOPulseBlockHandler():
+
+    def __init__(self, pb, assignment_dict, samp_rate=SEQ_SAMP_RATE, hd=None):
+        """ Initializes the pulse block handler for DIO
+
+        :hd: (object) And instance of the zi_hdawg.Driver()
+        :pb: (object) An instance of a pb.PulseBlock()
+        :samp_rate: (float) Sampling rate of HDAWG sequencer (300 MHz)
+
+        :assignment_dict: (dictionary) Dictionary mapping the channel names in the
+            pulse block to DIO channels. e.g.
+            {
+                'mw_gate' : 1,
+                'ctr' : 2
+            }
+            assigns the channel mw_gate to DIO bit 1, etc.
+        """
+
+        # Use the log client of the HDAWG
+        self.hd = hd
+        self.log = hd.log
+
+        # Store arguments.
+        self.pb = pb
+        self.assignment_dict = assignment_dict
+        self.DIO_bits = assignment_dict.values()
+
+        # Check key value integrity of assignment dict
+        self._check_key_assignments()
+
+        # Store remapped samples, number of samples and number of traces
+        self.sample_dict, self.num_samples, self.num_traces = self._get_remapped_samples(samp_rate=samp_rate)
+
+    def _check_key_assignments(self):
+        """Check if key values in assignment dict coincide with keys in pulseblock"""
+
+        if not self.assignment_dict.keys() == self.pb.p_dict.keys():
+            for assignment_key in self.assignment_dict.keys():
+                if assignment_key not in self.pb.p_dict.keys():
+                    self.log.error(
+                        f"Key '{assignment_key}' in assignment dictionary not found in pulseblock instance. Available keys are {self.pb.p_dict.keys()}."
+                    )
+            for pb_key in self.pb.p_dict.keys():
+                if pb_key not in self.assignment_dict.keys():
+                    self.log.error(
+                        f"Key '{pb_key}' in pulseblock instance not found in assignment dictionary."
```

Comment:
could be nice to give the user an option to manually assign a physical channel in the event that keys don't match

Reply:
Implemented in 765e1006f215f968c2cc49f13e83f817df0dbfa2. Now, the user can provide either an incomplete `assignment_dict` or none at all and will be asked to provide the missing DIO --> trace mappings.